File: Packages/StoryEngine/Sources/StoryEngine/DifficultyDirector.swift
import Foundation

public enum ActivityKind: String, Codable, Sendable {
    case navigation, puzzle, song, dialogue
}

public struct TelemetrySample: Codable, Equatable, Sendable {
    public var kind: ActivityKind
    public var succeeded: Bool
    public var duration: TimeInterval
    public var hintsUsed: Int

    public init(kind: ActivityKind, succeeded: Bool, duration: TimeInterval, hintsUsed: Int = 0) {
        self.kind = kind
        self.succeeded = succeeded
        self.duration = duration
        self.hintsUsed = hintsUsed
    }
}

/// How much help the game gives right now. All values move smoothly and
/// only ever change *support* — there is no punishment in this world.
public struct SupportLevel: Codable, Equatable, Sendable {
    /// Multiplier on guidance-sound volume, 1.0...2.0.
    public var audioCueGain: Double
    /// Multiplier on glow of important objects, 1.0...2.0.
    public var glowBoost: Double
    /// Index into each step's hintLadder: 0 gentle ... 2 fully explicit.
    public var hintTier: Int
    /// 1...5, drives song-spell length and puzzle richness.
    public var challenge: Int

    public static let standard = HintAggressiveness.standard.baseSupport

    public init(audioCueGain: Double, glowBoost: Double, hintTier: Int, challenge: Int) {
        self.audioCueGain = audioCueGain
        self.glowBoost = glowBoost
        self.hintTier = hintTier
        self.challenge = challenge
    }
}

/// Deterministic adaptive difficulty. Rules are intentionally legible:
/// predictability is itself an accessibility feature. An LLM may *phrase*
/// hints, but never decides how much help she gets. Support adapts relative
/// to the calibrated baseline and never drops below it.
public struct DifficultyDirector: Codable, Equatable, Sendable {
    public private(set) var support: SupportLevel
    public private(set) var baseline: SupportLevel
    private var recent: [TelemetrySample]
    private var windowSize: Int { 5 }

    /// A sample this slow counts as a struggle even when it succeeded.
    public var struggleDuration: TimeInterval

    public init(baseline: SupportLevel = .standard, struggleDuration: TimeInterval = 90) {
        self.baseline = baseline
        self.support = baseline
        self.struggleDuration = struggleDuration
        self.recent = []
    }

    public init(profile: CalibrationProfile) {
        self.init(baseline: profile.hintAggressiveness.baseSupport,
                  struggleDuration: profile.hintAggressiveness.struggleDuration)
    }

    /// Latest struggle-escalation, if record() just raised support — callers
    /// can log a struggleDetected event so the companion offers help.
    public private(set) var lastChangeWasEscalation: Bool = false

    public mutating func record(_ sample: TelemetrySample) {
        recent.append(sample)
        if recent.count > windowSize { recent.removeFirst(recent.count - windowSize) }
        adjust()
    }

    private func isStruggle(_ s: TelemetrySample) -> Bool {
        !s.succeeded || s.duration > struggleDuration || s.hintsUsed >= 2
    }

    private mutating func adjust() {
        let struggles = recent.suffix(2).filter(isStruggle).count
        let streak = recent.count >= 3 && recent.suffix(3).allSatisfy { !isStruggle($0) }
        lastChangeWasEscalation = false

        if struggles >= 2 {
            support.audioCueGain = min(2.0, support.audioCueGain + 0.2)
            support.glowBoost = min(2.0, support.glowBoost + 0.2)
            support.hintTier = min(2, support.hintTier + 1)
            support.challenge = max(1, support.challenge - 1)
            lastChangeWasEscalation = true
        } else if streak {
            support.audioCueGain = max(baseline.audioCueGain, support.audioCueGain - 0.1)
            support.glowBoost = max(baseline.glowBoost, support.glowBoost - 0.1)
            support.hintTier = max(baseline.hintTier, support.hintTier - 1)
            support.challenge = min(5, support.challenge + 1)
        }
    }

    /// The hint to speak for a step at the current support level, in the
    /// chosen companion's flavor.
    public func hint(for step: QuestStep, companionID: String?) -> String {
        guard !step.hintLadder.isEmpty else { return step.intro.resolved(for: companionID) }
        let index = min(support.hintTier, step.hintLadder.count - 1)
        return step.hintLadder[index].resolved(for: companionID)
    }
}

File: Packages/Lantern.Core/Runtime/Gameplay/DifficultyDirector.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.Linq;
using Lantern.Core.Content;

namespace Lantern.Core.Gameplay
{
    public enum ActivityKind { Navigation, Puzzle, Song, Dialogue }
    public sealed class TelemetrySample
    {
        public ActivityKind Kind { get; set; }
        public bool Succeeded { get; set; }
        public double Duration { get; set; }
        public int HintsUsed { get; set; }
    }
    public sealed class SupportLevel
    {
        public double AudioCueGain { get; set; } = 1.3;
        public double GlowBoost { get; set; } = 1.3;
        public int HintTier { get; set; } = 1;
        public int Challenge { get; set; } = 1;
        public static SupportLevel For(HintAggressiveness setting) => new SupportLevel
        { AudioCueGain = 1 + (int)setting * .3, GlowBoost = 1 + (int)setting * .3, HintTier = (int)setting };
    }
    public sealed class DifficultyDirector
    {
        public SupportLevel Support { get; set; } = new SupportLevel();
        public SupportLevel Baseline { get; set; } = new SupportLevel();
        public List<TelemetrySample> Recent { get; set; } = new List<TelemetrySample>();
        public double StruggleDuration { get; set; } = 90;
        public bool LastChangeWasEscalation { get; set; }
        public static DifficultyDirector For(CalibrationProfile profile) => new DifficultyDirector
        { Support = SupportLevel.For(profile.HintAggressiveness), Baseline = SupportLevel.For(profile.HintAggressiveness), StruggleDuration = 120 - 30 * (int)profile.HintAggressiveness };

        public void Record(TelemetrySample sample)
        {
            Recent.Add(sample);
            if (Recent.Count > 5) Recent.RemoveRange(0, Recent.Count - 5);
            LastChangeWasEscalation = false;
            if (Recent.Skip(Math.Max(0, Recent.Count - 2)).Count(IsStruggle) >= 2)
            {
                Support.AudioCueGain = Math.Min(2, Support.AudioCueGain + .2);
                Support.GlowBoost = Math.Min(2, Support.GlowBoost + .2);
                Support.HintTier = Math.Min(2, Support.HintTier + 1);
                Support.Challenge = Math.Max(1, Support.Challenge - 1);
                LastChangeWasEscalation = true;
            }
            else if (Recent.Count >= 3 && Recent.Skip(Recent.Count - 3).All(s => !IsStruggle(s)))
            {
                Support.AudioCueGain = Math.Max(Baseline.AudioCueGain, Support.AudioCueGain - .1);
                Support.GlowBoost = Math.Max(Baseline.GlowBoost, Support.GlowBoost - .1);
                Support.HintTier = Math.Max(Baseline.HintTier, Support.HintTier - 1);
                Support.Challenge = Math.Min(5, Support.Challenge + 1);
            }
        }
        private bool IsStruggle(TelemetrySample s) => !s.Succeeded || s.Duration > StruggleDuration || s.HintsUsed >= 2;
        public string Hint(QuestStep step, string? companionID) => step.HintLadder.Count == 0
            ? step.Intro.Resolve(companionID)
            : step.HintLadder[Math.Max(0, Math.Min(Support.HintTier, step.HintLadder.Count - 1))].Resolve(companionID);
    }
}
